sda/cmd/rotatekey/rotatekey.go

```diff
@@ -0,0 +1,268 @@
+	}
+
+	go func() {
+		connError := mq.ConnectionWatcher()
+		log.Error(connError)
+		forever <- false
+	}()
+
+	go func() {
+		connError := mq.ChannelWatcher()
+		log.Error(connError)
+		forever <- false
+	}()
+
+	log.Info("Starting rotatekey service")
+	var message schema.KeyRotation
+
+	go func() {
+		// Create a function to handle panic and exit gracefully
+		defer func() {
+			if err := recover(); err != nil {
+				if mq != nil {
+					defer mq.Channel.Close()
+					defer mq.Connection.Close()
+				}
+				if db != nil {
+					defer db.Close()
+				}
+				log.Fatal(err)
+			}
+		}()
+		messages, err := mq.GetMessages(conf.Broker.Queue)
+		if err != nil {
+			panic(err)
+		}
+		for delivered := range messages {
+			log.Debugf("Received a message (corr-id: %s, message: %s)",
+				delivered.CorrelationId,
+				delivered.Body)
+
+			err := schema.ValidateJSON(fmt.Sprintf("%s/rotate-key.json", conf.Broker.SchemasPath), delivered.Body)
+			if err != nil {
+				msg := "validation of incoming message (rotate-key) failed"
+				log.Errorf("%s, reason: %v", msg, err)
+				nackAndSendToErrorQueueckAndSendToErrorQueue(mq, delivered, conf.Broker.Exchange, msg, err.Error())
+
+				continue
+			}
+
+			// Fetch rotate key hash before starting work so that we make sure the hash state
+			// has not changed since the application startup.
+			keyhash := hex.EncodeToString(conf.RotateKey.PublicKey[:])
+			// exit app if target key was modified after app start-up, e.g. if key has been deprecated
+			if err = db.CheckKeyHash(keyhash); err != nil {
+				panic(fmt.Errorf("check of target key failed, reason: %v", err))
+			}
+
+			// we unmarshal the message in the validation step so this is safe to do
+			_ = json.Unmarshal(delivered.Body, &message)
+
+			fileID := message.FileID
+
+			// Get current keyhash for the file, send to error queue if this fails
+			oldKeyHash, err := db.GetKeyHash(fileID)
+			if err != nil {
+				msg := fmt.Sprintf("failed to get keyhash for file with file-id: %s", fileID)
+				log.Errorf("%s, reason: %v", msg, err)
+				nackAndSendToErrorQueueckAndSendToErrorQueue(mq, delivered, conf.Broker.Exchange, msg, err.Error())
+
+				continue
+			}
+
+			// Check that the file is not already encrypted with the target key
+			if oldKeyHash == keyhash {
+				log.Infof("the file with file-id: %s is already encrypted with the given rotation c4gh key", fileID)
+				if err := delivered.Ack(false); err != nil {
+					log.Errorf("failed to ack following already encrypted with key message")
+				}
+
+				continue
+			}
+
+			// reencrypt header
+			log.Debugf("rotating c4gh key for file with file-id: %s", fileID)
+
+			header, err := db.GetHeader(fileID)
+			if err != nil {
+				msg := fmt.Sprintf("GetHeader failed for file-id: %s", fileID)
+				log.Errorf("%s, reason: %v", msg, err)
+				nackAndSendToErrorQueueckAndSendToErrorQueue(mq, delivered, conf.Broker.Exchange, msg, err.Error())
+
+				continue
+			}
```

Comment:
same here we need to handle 2 different cases.

Reply:
The db function both for this and the above have a backoff strategy of tenths of seconds. It should be enough so that we can safely assume there is an actual error here. If we requeue, it may as well lead to infinite loops like we've seen in other services of the stack in recent past.